### PR TITLE
PERF: add shortcut to Timestamp constructor

### LINKED ISSUE #30543: PERF: Timestamp/Timedelta constructors when passed a Timestamp/Timedelta
The `Timestamp` constructor's performance could be improved when an existing `Timestamp` object is passed to it via an `isinstance`-like check:
```python
In [1]: import pandas as pd; pd.__version__
Out[1]: '0.26.0.dev0+1469.ge817ffff3'

In [2]: ts = pd.Timestamp('2020')

In [3]: def timestamp_isinstance_shortcircuit(ts): 
   ...:     if isinstance(ts, pd.Timestamp): 
   ...:         return ts 
   ...:     return pd.Timestamp(ts) 
   ...:

In [4]: %timeit pd.Timestamp(ts)
849 ns ± 13.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [5]: %timeit timestamp_isinstance_shortcircuit(ts)
121 ns ± 0.279 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
Some care is needed in the constructor to check if other arguments have been passed, e.g. `tz`, where we wouldn't be able to directly return the `Timestamp` object.

Similar story for the `Timedelta` constructor (should be done in a separate PR):
```python
In [6]: td = pd.Timedelta('1 day')

In [7]: def timedelta_isinstance_shortcircuit(td): 
   ...:     if isinstance(td, pd.Timedelta): 
   ...:         return td 
   ...:     return pd.Timedelta(td) 
   ...:

In [8]: %timeit pd.Timedelta(td)
800 ns ± 1.35 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [9]: %timeit timedelta_isinstance_shortcircuit(td)
120 ns ± 0.269 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

xref #30520

### PR DESCRIPTION
**Pull request by AlexKirko**
- [X] closes #30543 
- [X] tests added 1 / passed 1
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

This implements a shortcut in the Timestamp constructor to cut down on processing if Timestamp is passed. We still need to check if the timezone was passed correctly. Then, if a Timestamp was passed, and there is no timezone, we just return that same Timestamp.
A test is added to check that the Timestamp is still the same object.

PR for timedelta to be added once I confirm that this is the approach we want to go with.

### CODE REVIEW

**Comment by AlexKirko** [comment]
Simply adding a shortcut breaks date_range in an unintuitive way (see test `TestDatetimeIndexTimezones.test_dti_construction_ambiguous_endpoint`). I'll look deeper into this today to see why date_range no longer takes into account daylight savings with my fix.

File: pandas/_libs/tslibs/timestamps.pyx [hunk 1]
```diff
@@ -389,7 +389,10 @@ class Timestamp(_Timestamp):
             # User passed tzinfo instead of tz; avoid silently ignoring
             tz, tzinfo = tzinfo, None
 
-        if isinstance(ts_input, str):
+        if isinstance(ts_input, Timestamp) and tz is None:
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/timestamps.pyx]: you're going to need to check that all the other kwargs are None

Comment (AlexKirko) [reply pandas/_libs/tslibs/timestamps.pyx]: Thanks, I'll change this, but first I need to solve #24329, because right now we rely on Timestamp constructor changing ts.value when called on a localized DST Timestamp (this is why the tests for this PR break currently).

Comment (AlexKirko) [reply pandas/_libs/tslibs/timestamps.pyx]: Depending on how digging through the code goes, this PR might also take care of #24329, but I still need a couple days to analyze tz_localize, pd.date_range, and the Timestamp constructor.

File: pandas/tests/indexes/datetimes/test_timezones.py [hunk 2]
```diff
@@ -581,12 +581,7 @@ def test_dti_construction_ambiguous_endpoint(self, tz):
             ["US/Pacific", "shift_forward", "2019-03-10 03:00"],
             ["dateutil/US/Pacific", "shift_forward", "2019-03-10 03:00"],
             ["US/Pacific", "shift_backward", "2019-03-10 01:00"],
-            pytest.param(
-                "dateutil/US/Pacific",
-                "shift_backward",
-                "2019-03-10 01:00",
-                marks=pytest.mark.xfail(reason="GH 24329"),
-            ),
+            ["dateutil/US/Pacific", "shift_backward", "2019-03-10 01:00"],
```
Comment (AlexKirko) [inline pandas/tests/indexes/datetimes/test_timezones.py]: This shortcut, along with the fix to #24329 makes this exception no longer necessary, as a correct value gets returned without an error.

File: pandas/_libs/tslibs/timestamps.pyx [hunk 3]
```diff
@@ -379,6 +379,8 @@ class Timestamp(_Timestamp):
         _date_attributes = [year, month, day, hour, minute, second,
                             microsecond, nanosecond]
 
+        _non_ts_attributes = [freq, tz, unit, tzinfo] + _date_attributes
```
Comment (AlexKirko) [inline pandas/_libs/tslibs/timestamps.pyx]: I suggest we check for emtpy kwargs similar to the way we check for empty date attributes.

File: pandas/_libs/tslibs/timestamps.pyx [hunk 4]
```diff
@@ -391,7 +393,12 @@ class Timestamp(_Timestamp):
             # User passed tzinfo instead of tz; avoid silently ignoring
             tz, tzinfo = tzinfo, None
 
-        if isinstance(ts_input, str):
+        # GH 30543 if pd.Timestamp already passed, return it
+        # check that only ts_input is passed
+        if (isinstance(ts_input, Timestamp) and not
+                any(arg is not None for arg in _non_ts_attributes)):
```
Comment (AlexKirko) [inline pandas/_libs/tslibs/timestamps.pyx]: The same as with _date_attributes.

File: pandas/_libs/tslibs/timestamps.pyx (hunk 1 above)
Comment (AlexKirko) [reply pandas/_libs/tslibs/timestamps.pyx]: @jbrockmendel Solved #24329. Now this PR can be reviewed again.

**Comment by AlexKirko** [comment]
I suppose I should also move this to whatsnew for version 1.1.0?

File: pandas/_libs/tslibs/timestamps.pyx (hunk 4 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/timestamps.pyx]: This makes sense.  One question since we're focusing on performance: does it make a difference if you write out the verbose `and foo is None and bar is None and baz is None...`?  i have a suspicion that cython doesnt optimize this list comprehension

Comment (AlexKirko) [reply pandas/_libs/tslibs/timestamps.pyx]: @jbrockmendel Let's test it then.
The current implementation:
```
>>> ts = pd.Timestamp("2019-01-01")
>>> timeit.timeit(lambda: pd.Timestamp(ts), number=10000000)
5.860465700000002
```

And now with this implementation:
```
        if (isinstance(ts_input, Timestamp) and freq is None and
                tz is None and
                unit is None and
                year is None and
                month is None and
                day is None and
                hour is None and
                minute is None and
                second is None and
                microsecond is None and
                nanosecond is None and
                tzinfo is None):
            return ts_input
```

```
>>> ts = pd.Timestamp("2019-01-01")
>>> timeit.timeit(lambda: pd.Timestamp(ts), number=10000000)
4.43885800000001
```
You are right. Cython probably builds the entire list before starting boolean checks, so we incur on overhead.

File: pandas/_libs/tslibs/timestamps.pyx [hunk 5]
```diff
@@ -391,7 +391,16 @@ class Timestamp(_Timestamp):
...
 
-        if isinstance(ts_input, str):
+        # GH 30543 if pd.Timestamp already passed, return it
+        # check that only ts_input is passed
+        if (isinstance(ts_input, Timestamp) and freq is None and
+                tz is None and unit is None and year is None and
+                month is None and day is None and hour is None and
+                minute is None and second is None and
+                microsecond is None and nanosecond is None and
+                tzinfo is None):
```
Comment (AlexKirko) [inline pandas/_libs/tslibs/timestamps.pyx]: @jbrockmendel This appears the way to go if we need maximum performance.

Comment (WillAyd) [reply pandas/_libs/tslibs/timestamps.pyx]: I'm not sure all of these extra checks are worth adding to improve perf when passing a Timestamp to a Timestamp constructor - @AlexKirko how often are you expecting that to actually happen?

Comment (jbrockmendel) [reply pandas/_libs/tslibs/timestamps.pyx]: >  how often are you expecting that to actually happen?

For internal usage there are a lot of places where we do:

```
if isinstance(obj, (datetime, np.datetime64)):
    obj = Timestamp(obj)
```

That's not exactly the usage being checked here, but could benefit in the same way from an optimized no-other-arguments-passed check.

Comment (AlexKirko) [reply pandas/_libs/tslibs/timestamps.pyx]: Admittedly, I don't know the library well enough to comment on internal usage.

However, what I've done repeatedly in my own projects when on a deadline is take a Dataframe column or a list and just cast the elements into the type I need, trusting that if it already is that class, the performance loss won't be noticeable in the larger program (I do lots of ad-hoc data science modeling). I don't think it's as much of a problem in production-quality code, but a lot of people I work with use pandas to quickly preprocess data for sklearn.

You tend to rely on being able to cut corners when working with a well-supported package, and, currently, calling Timestamp on a Timestamp is about a 100 000 times slower than the proposed shortcut, which can be a nasty shock for someone expecting to just blitz through type conversions during data wrangling.

We could make the code a bit more compact with what was proposed in `4c9eb70`, which you can look up above, but this incurs about a 25% performance loss. I believe that if we do this, might as well add extra two lines. It's a bit grating but the gain is worth it, I think.

**Comment by jreback** [comment]
when you say cast? are you doing an astype or some sort of loop? can i show the original example perf issue

**Comment by AlexKirko** [comment]
When I say "cast" mean a `timeit` loop of `pd.Timestamp(ts)` in the console. Something along the lines of:
```
ts = pd.Timestamp(100)
timeit(lambda: pd.Timestamp(ts), number=10000000)
```
@jreback Thanks for suggesting running more testing. I went ahead and did that.

**TLDR:** depending on the usage conditions, the shortcut can speed up conversion to Timestamp to very different degrees. We speed things up about:
1. \>10 000 times in a `timeit` call in the console (testing done previously).
2. ~7 times in a `%timeit` call in Jupyter Notebook
3. ~30% when using `pd.Series.apply`
4. Nothing when using `pd.Series.astype`.

In addition:
1. Hoisting the check to the very top of `__new__`speeds up loop performance by about 20% and doesn't speed up anything else.
2. Replacing all the `is not None` checks with `not any` on a list comprehension slows down loop performance by about 100%, slows down `apply` a bit, and doesn't impact `astype`.

I think that implementing the shortcut is definitely worth it. As much as we might prefer it to not be true, some people will always loop when there is a better solution available (and sometimtes it's not available). I don't think moving the check to the very top of the function is worth it, as that is very ugly, but using expanded `is not None` gives a large performance boost in loops, so I think we should keep it.

What do you think?

The performance tests code and the results are below.

<details>

If we run the original perf issue code on master:
```
IN:
ts = pd.Timestamp('2020')
%timeit pd.Timestamp(ts)
OUT:
2.17 µs ± 168 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
And on the current commit of the PR branch (`bf08cc8`):
```
IN:
ts = pd.Timestamp('2020')
%timeit pd.Timestamp(ts)
OUT:
372 ns ± 31 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
So with original issue code, the shortcut runs about 7 times faster.
This is, of course, the worst possible case, but loops might be used only by somebody just starting with `pandas` and Python in general.

Let's test this further.
On master:
```
IN:
t_ser = pd.Series(range(100000))
t_ser = t_ser.astype('<M8[ns]')
%timeit t_ser.apply(lambda x: pd.Timestamp(x))
%timeit t_ser.astype('<M8[ns]')
OUT:
634 ms ± 25.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
146 µs ± 2.73 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
And on the current commit of the PR branch (`bf08cc8`):
```
IN:
t_ser = pd.Series(range(100000))
t_ser = t_ser.astype('<M8[ns]')
%timeit t_ser.apply(lambda x: pd.Timestamp(x))
%timeit t_ser.astype('<M8[ns]')
OUT:
464 ms ± 28.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
151 µs ± 4.07 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
So we save about 85% of time on a loop, 30% of time on `apply`, and nothing on `astype`. My guess is that `astype` has its own shortcut for when the user tries to cast a Series into its own dtype.
I also hoisted the check to the very top of the constructor and ran this again:
```
IN:
ts = pd.Timestamp('2020')
%timeit pd.Timestamp(ts)
OUT:
298 ns ± 18 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


IN:
t_ser = pd.Series(range(100000))
t_ser = t_ser.astype('<M8[ns]')
%timeit t_ser.apply(lambda x: pd.Timestamp(x))
%timeit t_ser.astype('<M8[ns]')
OUT:
451 ms ± 23 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
160 µs ± 8.24 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
So hoisting appears to help on loops.
And also ran it on commit `4c9eb70` (the solution with `not any(arg is not None for arg in _non_ts_attributes)`:
```
IN:
ts = pd.Timestamp('2020')
%timeit pd.Timestamp(ts)
OUT:
623 ns ± 25.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

IN:
t_ser = pd.Series(range(100000))
t_ser = t_ser.astype('<M8[ns]')
%timeit t_ser.apply(lambda x: pd.Timestamp(x))
%timeit t_ser.astype('<M8[ns]')
OUT:
482 ms ± 18.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
154 µs ± 3.65 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
With list comprehensions, this runs about two times slower in a loop, and a bit slower in lambda. My guess is that since the comprehension doesn't depend on the value converted, it gets optimized away.

</details>

Also pinging @jbrockmendel @WillAyd 
These testing results should make it easier to decide on what we want to implement here.

**Review by jreback — CHANGES_REQUESTED** [review]
@AlexKirko 

ok can you add the benchmarks you did as asv's.

I don't think this is really a big deal as converting scalars is a anti-pattern, we don't do this at scale (sure we do this often on a single basis), but on purpose we don't even store Timestamps (these are stored as i8), so if users are doing it and converting via ``.apply`` then tough.

that said this is ok as its not onerous on maitenance.

File: asv_bench/benchmarks/tslibs/timestamp.py [hunk 6]
```diff
@@ -28,6 +32,12 @@ def time_fromordinal(self):
     def time_fromtimestamp(self):
         Timestamp.fromtimestamp(1515448538)
 
+    def time_identity_scalar(self):
+        Timestamp(self.ts)
+
+    def time_identity_series_apply(self):
+        self.ts_series.apply(lambda x: Timestamp(x))
```
Comment (AlexKirko) [inline asv_bench/benchmarks/tslibs/timestamp.py]: @jreback 

Added benchmarks for scalar cast and apply. Didn't touch astype, as that has nothing to do with this PR. Is this what you had in mind?

Comment (jbrockmendel) [reply asv_bench/benchmarks/tslibs/timestamp.py]: benchmark.tslibs is intended to only depend on _libs.tslibs, so shouldnt have any Series objects in it.  just benchmark the Timestamp contructor here

Comment (jbrockmendel) [reply asv_bench/benchmarks/tslibs/timestamp.py]: as long as we're at it, might as well have cases for single-arguments for np.datetime64, pydatetime, and tzaware

File: pandas/_libs/tslibs/timestamps.pyx (hunk 5 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/timestamps.pyx]: can you add a comment to the effect of "we do this verbose thing because cython wont optimize a list comprehension (as of cython 0.29.x)"

File: asv_bench/benchmarks/tslibs/timestamp.py [hunk 7]
```diff
@@ -28,6 +35,18 @@ def time_fromordinal(self):
...
+        Timestamp(self.npdatetime64)
+
+    def time_from_datetime_unaware(self):
+        Timestamp(self.dttime_unaware)
+
+    def time_from_datetime_aware(self):
+        Timestamp(self.dttime_aware)
+
+    def time_from_pd_timestamp(self):
+        Timestamp(self.ts)
```
Comment (AlexKirko) [inline asv_bench/benchmarks/tslibs/timestamp.py]: @jbrockmendel 
Was this what you had in mind? I think we should leave only the scalar transformer for benchmarking this shortcut. Don't see much sense in adding the benchmarks for transforming a Series.

Comment (jreback) [reply asv_bench/benchmarks/tslibs/timestamp.py]: this looks good

**Review by jreback — APPROVED** [review]
lgtm. i think if you add @jbrockmendel comment about cython should be good. ping on green.

**Comment by AlexKirko** [comment]
@jreback 
Done.

File: asv_bench/benchmarks/tslibs/timestamp.py (hunk 7 above)
Comment (jbrockmendel) [reply asv_bench/benchmarks/tslibs/timestamp.py]: perfect, thanks

File: pandas/_libs/tslibs/timestamps.pyx (hunk 5 above)
Comment (AlexKirko) [reply pandas/_libs/tslibs/timestamps.pyx]: @jbrockmendel Added the comment you requested.

**Comment by AlexKirko** [comment]
@jreback 
All green, should be ready to merge.

**Comment by jreback** [comment]
lgtm. can you rebase, ping on green.

**Comment by AlexKirko** [comment]
@jreback 
Rebased, all green.
Also cleaned up the conflicts with my previour 2 PRs. Don't need one of the `dateutil` xfails now.

**Comment by jreback** [comment]
thanks @AlexKirko very nice! keep em coming
